### PR TITLE
Yaml driven api

### PR DESCRIPTION
**Pull request by paddyzab**
To make the api more flexible we are storing the map of the interested assets in form of a yaml map.

Writing and reading this representation to the cache.

Next planned step to decorate the results printed to the console with information about the bags used owned by the user - again thanks to the yaml file.

### CODE REVIEW

File: pkg/cointicker/cointicker.go [hunk 1]
```diff
@@ -42,8 +42,11 @@ func (c CoinTicker) GetFormattedPrice(t time.Time) (string, []error) {
 
 	fmt.Println(res)
 
-	//lastEntry := c.Cache.GetLast()
-	//c.Cache.AddEntry(btc, eth, xmr, neo, float.Round(ethRatio), float.Round(xmrRatio), float.Round(neoRatio), t.UTC())
+	lastEntry := c.Cache.GetLast()
+	c.Cache.AddEntry(res, t.UTC())
+
+	fmt.Print("last entry:")
```
Comment (alagsa) [inline pkg/cointicker/cointicker.go]: You could collapse this two prints into one.

File: pkg/storage/storage.go [hunk 2]
```diff
@@ -1,64 +1,61 @@
...
-		MoneroPrice:  m,
-		NeoPrice:     n,
-		ETHRatio:     ber,
-		XMRRatio:     mer,
-		NEORatio:     ner,
-		Timestamp:    t,
+		CoinData:  res,
+		Timestamp: t,
 	})
+	fmt.Print("adding to cache: ")
```
Comment (alagsa) [inline pkg/storage/storage.go]: Do you really want to keep this?

File: pkg/cointicker/cointicker.go [hunk 3]
```diff
@@ -35,16 +34,22 @@ func NewCoinTicker(client *cmcap.CoinMarketClient, cache *storage.Cache) CoinTic
...
 		return "", errors
 	}
 
+	fmt.Println(res)
+
 	lastEntry := c.Cache.GetLast()
-	c.Cache.AddEntry(btc, eth, xmr, neo, float.Round(ethRatio), float.Round(xmrRatio), float.Round(neoRatio), t.UTC())
+	c.Cache.AddEntry(res, t.UTC())
+
+	fmt.Print("last entry:")
```
Comment (alagsa) [inline pkg/cointicker/cointicker.go]: Do you want to keep this?

File: pkg/cointicker/cointicker.go [hunk 4]
```diff
@@ -35,16 +34,22 @@ func NewCoinTicker(client *cmcap.CoinMarketClient, cache *storage.Cache) CoinTic
...
+	fmt.Println(lastEntry)
 
-	return fmt.Sprintf("%s BTC: %s, ETH: %s, XMR: %s, NEO: %s \nB/E ratio %f, B/M ratio %f, B/N ratio %f \n\n", t.Format(timeFormat), btc, eth, xmr, neo, decorateRatio(ethRatio, lastEntry.ETHRatio, c)(ethRatio), decorateRatio(xmrRatio, lastEntry.XMRRatio, c)(xmrRatio), decorateRatio(neoRatio, lastEntry.NEORatio, c)(neoRatio)), nil
+	//return fmt.Sprintf("%s BTC: %s, ETH: %s, XMR: %s, NEO: %s \nB/E ratio %f, B/M ratio %f, B/N ratio %f \n\n", t.Format(timeFormat), btc, eth, xmr, neo, decorateRatio(ethRatio, lastEntry.ETHRatio, c)(ethRatio), decorateRatio(xmrRatio, lastEntry.XMRRatio, c)(xmrRatio), decorateRatio(neoRatio, lastEntry.NEORatio, c)(neoRatio)), nil
```
Comment (alagsa) [inline pkg/cointicker/cointicker.go]: 3 consecutive `return` statements  :)

File: pkg/cointicker/cointicker.go [hunk 5]
```diff
@@ -55,23 +60,25 @@ func decorateRatio(r, lr float64, c CoinTicker) func(interface{}) aurora.Value {
 	return c.au.Red
 }
 
-// Results ...
-type Results struct {
-	Result map[string]float64
-}
+func (c CoinTicker) generateResult() (results storage.Results, errors []error) {
```
Comment (alagsa) [inline pkg/cointicker/cointicker.go]: Maybe you could not use named returns here and create a `Results` element from scratch (perhaps even using a `NewResults` type of function) in the method.

Comment (alagsa) [reply pkg/cointicker/cointicker.go]: Also returning from `cointicker` something that is defined in `storage` may not be so untangled as you'd like.

**Review by alagsa — APPROVED** [review]
In general I think what you want to achieve here is to have a yaml that allows for further addition of new coins (as you expand your portfolio) without code changes. This is great and I like the yaml approach.

There are only a couple of things regarding where some types are defined (storage may not be the best place for a cross-package dependency). Other than that, it seems to be a step forward!

File: pkg/coinmarketcap/coin.go [hunk 6]
```diff
@@ -17,20 +17,10 @@ type Coin struct {
 	LastUpdated      string `json:"last_updated"`
 }
 
-// Coins represents a collection of Coin
-type Coins []Coin
+// CoinsMap represents a map of coins with the key being Coin#Symbol of a Coin
+type CoinsMap map[string]Coin
 
 // Len to implement the Sort interface
```
[the remark this answers is not in the document]
Comment (paddyzab) [reply pkg/coinmarketcap/coin.go]: removing, was used only for sorting interface.

**Pull request merged**